src/lsquic/client.cr

```diff
@@ -213,17 +213,24 @@ module QUIC
         end
         @engine_open = false
         LibLsquic.engine_destroy(engine)
+        @socket.try &.close
+        @socket = nil
       end
 
-      buffer = Bytes.new(0x600)
-      loop do
-        bytes_read = socket.read buffer
-        break if !@engine_open
-        LibLsquic.engine_packet_in(engine, buffer[0, bytes_read], bytes_read, socket.local_address, socket.remote_address, Box.box(socket), 0) if bytes_read != 0
-        LibLsquic.engine_process_conns(engine)
+      begin
+        buffer = Bytes.new(0x600)
+        loop do
+          bytes_read = socket.read buffer
+          break if !@engine_open
+          LibLsquic.engine_packet_in(engine, buffer[0, bytes_read], bytes_read, socket.local_address, socket.remote_address, Box.box(socket), 0) if bytes_read != 0
+          LibLsquic.engine_process_conns(engine)
+        end
+      rescue IO::Error
+        # may have already been closed
+      ensure
+        @socket.try &.close
+        @socket = nil
```

Comment:
To give the context the context that I found while reviewing this...

This method, `run_engine` is spawned whenever a request is made if it hasn't already been spawned. So, if an error occurred, the socket wasn't closed and the whole thing was made over for the next request. By adding the ensure part, we make sure that the socket is closed even on error.

Reply:
The ensure may not be necessary then and only close socket if the engine has been destroyed.